src/boot/efi/vmm.c

```diff
@@ -289,3 +295,21 @@ bool in_hypervisor(void) {
         cache = cpuid_in_hypervisor() || smbios_in_hypervisor();
         return cache;
 }
+
+const char* smbios_find_oem_string(const char *name) {
+        assert(name);
+
+        SmbiosTableType11 *type11 = (SmbiosTableType11 *) get_smbios_table(11);
+        if (!type11 || type11->header.length < sizeof(SmbiosTableType11))
+                return NULL;
+
+        NULSTR_FOREACH8(s, type11->contents) {
```

Comment:
NULSTR_FOREACH8() is not the right choice for cases like this that have a length. i.e. NULSTR_FOREACH8() relies that the nulstr ends in two NUL bytes. But that#s not a given here.

This really should be more defensive, and strictly take the length into account, and even be fine if the buffer doesn't end with a NUL at all...

Reply:
The SMBIOS spec mandates that the list of strings ends with 2 nul bytes so this should be fine

Reply:
Does the spec also account for vendors not fucking things up?

Reply:
this sounds like something vendors will very very likely fuck up, since the double NUL thing is redundant, and I am pretty sure many OSes will just safely handle because the size exists. Ignoring the size here is a major mistake to me.

Reply:
The length field only specifies the length of the formatted region of the struct, without the strings at the end. We **have** to use the NUL bytes to detect the end of the strings.

Reply:
The length of a smbios table only specifies the length of the formatted region and does not include the strings at the end. The only way to detect the end of the strings is to use the double NUL byte, you can't use the length of the table.

Reply:
SmbiosHeader has a length field, no?

get_smbios_table() probably should return that in a second, optional argument so that we can use it here.

Reply:
The smbiosheader length field does not include the length of the strings at the end.

Reply:
right, but find_smbios_configuration_table() returns the length. Got mixed up in all those tables...

the length that function returns should be taken into account here.

So I guess get_smbios_table() should be updated to optionally return the maximum size of the table that was found. (i.e. get a `ret_size_left` param, that just returns how much bytes are not read yet of the overall SMBIOS table)